GWFish/modules/waveforms.py

```diff
@@ -113,51 +130,308 @@ def t_of_f_PN(parameters, frequencyvector):
+                self.delta_f,
+                self.f_min,
+                self.f_max,
+                self.f_ref,
+                self._params_lal,
+                self._approx_lal
+            ]
+            logging.warning("{} does not allow arbitrary frequency vectors "
+                            "in LALSimulation, using standard frequency vector."
+                            "Try calling this waveform outside lalsim.".format(self.name))
+            logging.warning("Parameters phiRef, longAscNodes, eccentricity, meanPerAno"
+                            "are set to zero.")
+
+    def _update_frequency_range_indices(self):
+        self.idx_low = int(self.f_min / self.delta_f)
+        self.idx_high = int(self.f_max / self.delta_f)
+
+    def _lal_fd_strain_adjust_frequency_range(self):
+        """ Frequency array starts from zero, so we need to mask some frequencies """
+        self._update_frequency_range_indices()
+        self.hf_cross_out = self._lal_hf_cross.data.data[self.idx_low:self.idx_high+1]
+        self.hf_plus_out = self._lal_hf_plus.data.data[self.idx_low:self.idx_high+1]
+
+    def _lal_fd_phase_correction_by_epoch_and_df(self):
+        """ This correction is also done in Bilby after calling SimInspiralFD """
+        # BORIS: weird Bilby correction
+        dt = 1. / self.delta_f + (self._lal_hf_plus.epoch.gpsSeconds +
+                                  self._lal_hf_plus.epoch.gpsNanoSeconds * 1e-9)
+        self.hf_plus_out *= np.exp(
+            -1j * 2 * np.pi * dt * self.frequencyvector)
+        self.hf_cross_out *= np.exp(
+            -1j * 2 * np.pi * dt * self.frequencyvector)
+
+    def _hf_postproccessing_SimInspiralFD(self):
+        self._lal_fd_strain_adjust_frequency_range()
+        self._lal_fd_phase_correction_by_epoch_and_df()
+
+    def _hf_postproccessing_SimInspiralCFDWS(self):
+        self.hf_plus_out, self.hf_cross_out = self._lal_hf_plus.data.data, self._lal_hf_cross.data.data
+
+    def _fd_phase_correction_geocent_time(self):
+        """ Add initial 2pi*f*tc - phic - pi/4 to phase """
+        phi_in = np.exp(1.j*(2*self.frequencyvector*np.pi*self.gw_params['geocent_time']))
+
+        hfp = phi_in * np.conjugate(self.hf_plus_out)  # it's already multiplied by the phase
+        hfc = phi_in * np.conjugate(self.hf_cross_out)
+
+        return hfp, hfc
+
+    def _fd_gwfish_output_format(self, hfp, hfc):
+
+        hfp = hfp[:, np.newaxis]
+        hfc = hfc[:, np.newaxis]
+
+        polarizations = np.hstack((hfp, hfc))
+
+        print('Warning: inverting hx in LAL caller')
+
+        return polarizations
+
+    def frequency_domain_strain(self):
+        self._lal_hf_plus, self._lal_hf_cross = self._lalsim_caller(*self._lalsim_args)
+        self._waveform_postprocessing()
+
+        hfp, hfc = self._fd_phase_correction_geocent_time()
+        polarizations = self._fd_gwfish_output_format(hfp, hfc)
+        
+        return polarizations
+
+class LALTD_Waveform(LALFD_Waveform):
```

Comment:
I think this name is maybe misleading - the waveforms you get when calling this object are still FD, but they come from the TD - maybe `LALFD_Waveform_fromTD`?

Reply:
I see your point, but `FD_fromTD` is also not perfect because the object contains both TD and FD waveform, TD part is not completely discarded. I resolved this by adding more informed docstrings to both classes, they should better explain things.